src/util/VariableNameCreator.ts

```diff
@@ -17,69 +19,126 @@ const reservedKeywords = [
 ];
 const sapReservedKeywords = [ "sap" ];
 
+const reservedBrowserTypes = Object.keys(globals.browser);
+const reservedNativeTypes = Object.keys(globals.builtin);
 
-export function getUniqueVariableName(aUsedVariables: string[], sName: string) {
-	let sResultName = "";
-	sName.split(".").reverse().some((sNamePart) => {
-		sResultName = decapitalize(sNamePart) + capitalize(sResultName);
-		if (!aUsedVariables.includes(normalize(sResultName))) {
-			sResultName = normalize(sResultName);
-			return true;
-		}
-		return false;
-	});
+export function getUniqueParameterName(
+	aUsedVariables: string[], sName: string) {
+	return getUniqueName(aUsedVariables, sName);
+}
 
-	while (aUsedVariables.includes(sResultName)) {
-		sResultName = "o" + normalize(capitalize(sResultName));
-	}
+const getUniqueNameDecapitalized = function(
+	aUsedVariables: string[], sName: string) {
+	return getUniqueName(aUsedVariables, sName, true);
+};
 
-	return sResultName;
-}
+const isLowerCase = function(sChar) {
+	return sChar === sChar.toLowerCase();
+};
 
 /**
  *
- * @param sVariableName the input variable name
- * @returns variable name which neither contain invalid characters nor is a
- * language keyword
+ * @param aUsedVariables
+ * @param sName
  */
-export function normalize(sVariableName: string): string {
-	// variable name must start with a letter or an underscore
-	// if not prepend "s" for string
-	if (!rAllowedStartCharacter.test(sVariableName)) {
-		sVariableName = "o" + capitalize(sVariableName);
-	}
+const isUnique = function(aUsedVariables: string[], sName: string): boolean {
+	return !aUsedVariables.includes(sName) && !isReservedWord(sName);
+};
 
-	// prettify variable name by removing dashes
-	if (sVariableName.includes("-")) {
-		const rCamelCase = /-(.)/ig;
-		sVariableName =
-			sVariableName.replace(rCamelCase, function(sMatch, sChar) {
-				return sChar.toUpperCase();
-			});
-	}
+const isValidIdentifierName = function(
+	aUsedVariables: string[], sName: string) {
+	return rAllowedStartCharacter.test(sName) &&
+		isUnique(aUsedVariables, sName);
+};
+
+const replaceInvalidCharacters = function(sName) {
+	sName = camelize(sName);
+	return sName.replace(rInvalidChars, "_");
+};
 
-	// replace invalid chars
-	sVariableName = sVariableName.replace(rInvalidChars, "_");
 
-	// ensure there is no keyword match
-	while (reservedKeywords.includes(sVariableName) ||
-		   sapReservedKeywords.includes(sVariableName)) {
-		sVariableName = "o" + capitalize(sVariableName);
+const getUniqueName = function(
```

Comment:
Please add jsdoc, this would help to understand the prefixing

Reply:
Only when the whole namespace is still no valid variable name, the prefixing is taking place

Reply:
ok